doc: add policy for distribution (#51918)
nodejs/node: doc/contributing/distribution.md

# Node.js Distribution Policy

This document describes some policies around what is and is not included in the
Node.js distribution.

## Inclusion

Node.js includes many dependencies, such as V8, Undici and others; and comes
bundled with standalone tools such as `npm`. It is not a goal of the Node.js

> project to provide a level playing field for all dependencies or bundled tools.

Reviewer: Instead of listing non-goals, let’s say what you really mean here: npm distribution in the default installer is only justified by historical reasons, and is likely going to stay this way for the foreseeable future to avoid disturbing the ecosystem.

Author: I don’t know why you think that that’s what I really mean. I don’t view npm’s inclusion as some kind of regretful historical artifact. It’s the standard package manager, for no other reason than that it’s published by the people who run the de facto standard package registry, and so therefore its inclusion is a sensible default.

The point of this is so that when someone complains that it’s not fair that we ship npm, so for equity/fairness/level playing field/competition/anticompetitiveness/etc therefore we need to also ship one or more other package managers, the reason why we don’t is because our policy is to ship only one type of each category of tool. It’s a drain on finite resources to support multiple, and promoting fairness in the marketplace is not our responsibility.

Reviewer 2: I believe this is trying to answer the question I was trying to ask in #50963 which **is** a more general question. I think that from the perspective of attempting to "provide clarity" this more general wording is better than the more direct thing even if we all know the reason. This way the document presents a more general stance of the project.

I do think that the "non-goals" framing is maybe not necessary to get the right point across. The point is to follow the [technical values](https://github.com/nodejs/node/blob/f4af4b111c20e599742223bc1a78b0a7017870a7/doc/contributing/technical-values.md) and put DX and stability first. Corepack is not the best we as a community can do for DX and knowing that, enabling it by default is not the best solution is not good for future stability. Maybe a better way to state this is:

```suggestion
It is the goal of the Node.js project to provide a great and stable developer experience, which means we cannot ship all things for all user groups.
```

Reviewer 2: Oh this was marked resolved as I commented. @GeoffreyBooth maybe could we re-open it until the conversation is complete?

Author: This wasn’t inspired by Corepack, it was inspired the previous discussion around shipping Yarn: https://github.com/nodejs/TSC/issues/1012#issuecomment-820476633. I really meant what I wrote here, that in general we shouldn’t ship multiple tools for the same purpose. That’s not fair to Yarn, yes, but there are only so many Node contributors and promoting competition in the ecosystem is not one of our project goals (not that we oppose competition or fairness by any means, but it shouldn’t come at a cost of draining time and effort from the Node project itself).

Reviewer 3: > Instead of listing non-goals, let’s say what you really mean here: npm distribution in the default installer is only justified by historical reasons, and is likely going to stay this way for the foreseeable future to avoid disturbing the ecosystem.

I would encourage you not to presume intent when the justification was explicitly written (ie. `"It is not a goal of the Node.js
project to provide a level playing field for all dependencies or bundled tools."`). There is no reference to historical precedence in this PR. 

As @GeoffreyBooth has already confirmed, this policy simply codifies that `"a level playing field"` is not criteria or justification that will be considered when determining a dependency or bundled tool's value to the project.

I'm fine with the wording. A "non-goal" is simply a "statement" & statements are required to define a "policy". If its accepted, & merged, it becomes "fact".

Reviewer 2: That makes sense, I didn't mean to make any comments on your intent, just that it is alright if folks think (know) that this new policy is required *because* of all the other things. That doesn't change the longer term value in having clarity on this topic, nor does it mean the wording is/is not good.  think that this wording is framing as "negative" what could be phrased as a "positive".  This really is about what the project **is doing** (providing good dx and stability for the ecosystem).

Author: I mean, I could just delete “It is not a goal of the Node.js project to provide a level playing field for all dependencies or bundled tools.” if it bothers people. It’s arguably redundant given the following sentence.

The problem with saying that a goal of Node is to promote good DX is a) we have a different document that defines our technical values, and b) good DX is in the eye of the beholder, and there are those who would argue that a vibrant ecosystem is good DX, and therefore a goal of good DX commits us to a competitive ecosystem etc. What I _intend_ it to close off a justification for including potential additions to our distribution, and the most efficient way to phrase that (that I can think of) is to state non-goals.

Reviewer: I think that sentence should be removed, listing things which are "not a goal" is a bad idea.

> I don’t know why you think that that’s what I really mean.

Sorry for misreading you then.

Reviewer 4: > I think that sentence should be removed, listing things which are "not a goal" is a bad idea.

I explicitly disagree. When referring to a document like this, understanding what the scope is _intended_ to be used for and what it's _not intended_ to be used for is extremely valuable.

Reviewer: In this case, I'd challenge that it'd be extremely valuable. Please remove it.

Reviewer 2: Note: sorry for the duplicate, in an attempt to fix the broken GH UI from submitting the comment with a review approval my original comment was deleted. Trying to recreate it here.

I think there is a better way to get the same point across without the negative framing. Something along the lines of:

> It is the goal of the Node.js project to provide a stable UX and this means that we cannot ship all features/tools/etc for all user groups.

With limited resources, the Node.js project does not intend to bundle or
officially support multiple dependencies or tools that serve the same purpose.

> ## Package managers

Reviewer 5: Should we add dependencies to this list?

Author: As in, add more sections? Sure, we can, but I figured that this document can always expand over time.

Reviewer 4: Might be worth at least documenting the other ones mentioned earlier in the document - V8 and Undici.

> The Node.js distribution will include `npm`, and will not include any other

Reviewer: ```suggestion
The Node.js distribution includes `npm`, and does not include any other
```

> package managers.

Reviewer 6: ```suggestion
package managers.
We bundle npm because it's the _reference implementation_ of the npm registry,
as well as historical reasons.
```

Author: I'm not opposed to this, but if we're also adding sections for V8 and Undici then I'll need to also explain why each of those were chosen, and I don't really know. Perhaps we can leave out the history for now and it can be added in a follow up?

Reviewer 6: I think it would be good to document those things. I can add a paragraph about undici (mostly because it's maintained by the same people who maintain the node core http client).

Reviewer 2: I don't think we need to leave out the history, but also we could just say something like:

> The decision was made many years before alternatives exist, and unless there is strong reason to switch away from the chosen solution the decision will not be regularly revisited.

Reviewer 7: I think we could start by explaining those that people can add a suggestion with the explanation (as Matteo volunteered for undici) and just leave the others blank until if/when sombody PRs in an explanation.

# Node.js Distribution Policy

This document describes some policies around what is and is not included in the
Node.js distribution.

## Inclusion

> Node.js includes some external projects that the Node.js team does not maintain.

Author: I’m referring to the Node.js distribution, not the repo. Should this say “The Node.js distribution”? I assumed it was obvious because the entire document is about the Node.js distribution.

> Node.js includes some external projects that the Node.js team does not maintain.

Reviewer: Maybe say "it" if you don't want to repeat it
```suggestion
It includes some external projects that the Node.js project does not maintain.
```

> Node.js includes some external projects that the Node.js team does not maintain.

Author: I don’t think we should start a section with “It.” Is that what you’re looking for?

```suggestion
The Node.js distribution includes some external projects that the Node.js project does not maintain.
```

> Node.js includes some external projects that the Node.js team does not maintain.

Reviewer: That would work yes. Another idea would be to use "Official distributions", no string feelings though

> Node.js includes some external projects that the Node.js team does not maintain.

Author: Saying “official distributions” implies that maybe there are other policies for “unofficial” distributions, but I don’t want to open up that can of worms.

The fact of a project's inclusion should not imply anything about the project

> relative to its competitors; in some cases, a project was added when it had no

Author: This was to address @mcollina’s note about how some choices of what to include were made for historical reasons.

> relative to its competitors; in some cases, a project was added when it had no

Reviewer: Maybe it would help if we find a different word than project, and only the word "project" to refer to the Node.js project. Maybe "dependency" would work?

> The fact of a project's inclusion should not imply anything about the project
> relative to its competitors; in some cases, a project was added when it had no

Author: The problem with “dependency” is that npm isn’t really a dependency of Node, nor is Corepack. They’re just additional pieces of software we happen to include as part of our distribution. Perhaps this?

```suggestion
The fact of a tool or library's inclusion should not imply anything about it
relative to its competitors; in some cases, something was added when it had no
```

> competitors. While the Node.js team supports and encourages competition in the

Reviewer: ```suggestion
competitors. While the Node.js project supports and encourages competition in the
```

JavaScript ecosystem, as a policy, the Node.js project does not include multiple
dependencies or tools that serve the same purpose.

The following user-accessible external projects are the ones chosen for their
particular purposes:

* JavaScript engine: V8

> * Package manager: `npm`

Reviewer: ```suggestion
* Package manager: `npm`
* Package manager manager: Corepack
* `fetch` implementation: Undici
```

Author: I think Corepack is more properly described as a package manager version manager.

Undici isn’t exposed to users, is it? I was only listing the visible ones.

Reviewer: I'd argue it is, via some "global symbols"

Author: ```suggestion
* Package manager: `npm`
* Package manager version manager: Corepack
```
